sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java

```diff
@@ -0,0 +1,633 @@
+import com.google.cloud.firestore.v1.FirestoreClient.PartitionQueryPage;
+import com.google.cloud.firestore.v1.FirestoreClient.PartitionQueryPagedResponse;
+import com.google.cloud.firestore.v1.stub.FirestoreStub;
+import com.google.firestore.v1.BatchGetDocumentsRequest;
+import com.google.firestore.v1.BatchGetDocumentsResponse;
+import com.google.firestore.v1.Cursor;
+import com.google.firestore.v1.ListCollectionIdsRequest;
+import com.google.firestore.v1.ListCollectionIdsResponse;
+import com.google.firestore.v1.ListDocumentsRequest;
+import com.google.firestore.v1.ListDocumentsResponse;
+import com.google.firestore.v1.PartitionQueryRequest;
+import com.google.firestore.v1.PartitionQueryResponse;
+import com.google.firestore.v1.RunQueryRequest;
+import com.google.firestore.v1.RunQueryResponse;
+import com.google.firestore.v1.StructuredQuery;
+import com.google.firestore.v1.StructuredQuery.Direction;
+import com.google.firestore.v1.StructuredQuery.FieldReference;
+import com.google.firestore.v1.StructuredQuery.Order;
+import com.google.firestore.v1.Value;
+import com.google.protobuf.Message;
+import com.google.protobuf.ProtocolStringList;
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.beam.sdk.extensions.gcp.options.GcpOptions;
+import org.apache.beam.sdk.io.gcp.firestore.FirestoreDoFn.NonWindowAwareDoFn;
+import org.apache.beam.sdk.io.gcp.firestore.FirestoreV1Fn.HasRpcAttemptContext;
+import org.apache.beam.sdk.io.gcp.firestore.RpcQos.RpcAttempt.Context;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.checkerframework.checker.nullness.compatqual.NullableDecl;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.joda.time.Instant;
+
+/**
+ * A collection of {@link org.apache.beam.sdk.transforms.DoFn DoFn}s for each of the supported read
+ * RPC methods from the Cloud Firestore V1 API.
+ */
+final class FirestoreV1ReadFn {
+
+  /**
+   * {@link DoFn} for Firestore V1 {@link RunQueryRequest}s.
+   *
+   * <p>This Fn uses a stream to obtain responses, each response from the stream will be output to
+   * the next stage of the pipeline. Each response from the stream represents an individual document
+   * with the associated metadata.
+   *
+   * <p>If an error is encountered while reading from the stream, the stream will attempt to resume
+   * rather than starting over. The restarting of the stream will continue within the scope of the
+   * completion of the request (meaning any possibility of resumption is contingent upon an attempt
+   * being available in the Qos budget).
+   *
+   * <p>All request quality-of-service is managed via the instance of {@link RpcQos} associated with
+   * the lifecycle of this Fn.
+   */
+  static final class RunQueryFn
```

Comment:
Beam sources help Beam pipeline parallelize. I think we have to update following source Fn classes to use SDF to parallelize better. This help support features such as dynamic work rebalancing. Without proper parallelization Beam pipelines that use Firestore source could run into stragglers (which is an issue many Dataflow customers run into without dynamic work rebalancing is not available).
See here for more details on SDF: https://beam.apache.org/documentation/programming-guide/#splittable-dofns

Reply:
I would also try to reduce the the implementation to one or few SDF-based sources to make implementation simpler.

Reply:
I did attempt to provide a SDF for query when implementing this class, however today the API for `PartitionQuery` does not allow dynamic splitting, it is a purely pre-process type of operation. Due to this fact, I had to add the additional `PartitionQueryFn` which could perform the pre-processing and yield the computed queries.

Reply:
@nehsyc do you have any concerns here related to Dataflow autoscaling ?

Reply:
Limited parallelism (by the number of partitions) would be the main concern w.r.t. performance, especially when the processing time for each partitioned query can vary a lot. If the static partitioning done by Datastore is fairly good in terms of balanced load breakdown, dynamic sharding would become less needed there. It'd be good anyway to document the known drawback of the manual partitioning.

Once Dataflow detects the limited parallelism, it will not upsize more than that to avoid wasting resources. All those happen automatically when Dataflow fails splitting the tasks (or bundles) for a short period of time (minutes).

BTW Ben mentioned another point related to autoscaling, throttling during the read. Read might fail and get retried which could cause Dataflow to upsize the worker pool. Reporting throttling signals like what's done in the sink back to Dataflow would potentially solve the problem.

Reply:
On both the read and write path, if the workers is throttling for some backoff it is reported via `Metrics.counter("throttlingMs").inc(throttleDuration.getMillis())`. In the case of read, the only time a throttle would would occur is if RPCs are failing and backoff kicks in. In the case of write, in addition or RPC failure backoff, a worker can be throttled as part of write ramp up, or if RPC are not failing, but writes within those RPCs are failing and the client side adaptive throttler kicks in to try and bring the error rate down.

Reply:
Ben could you add a TODO with a Jira to support progress reporting and dynamic work rebalancing in the future ?

Reply:
Done, and added as TODOs in on the class
* BEAM-12605 Add dynamic work rebalancing to Firestore Query Source
* BEAM-12606 Add support for progress reporting to Firestore Query Source